src/async-wrap.cc

```diff
@@ -225,35 +372,46 @@ Local<Value> AsyncWrap::MakeCallback(const Local<Function> cb,
     }
   }
 
-  if (ran_init_callback() && !pre_fn.IsEmpty()) {
+  if (async_hooks->fields()[Environment::AsyncHooks::kBefore] > 0) {
+    uid = Number::New(env()->isolate(), get_uid());
+    Local<Function> fn = env()->async_hooks_before_function();
     TryCatch try_catch(env()->isolate());
-    MaybeLocal<Value> ar = pre_fn->Call(env()->context(), context, 1, &uid);
+    MaybeLocal<Value> ar = fn->Call(
+        env()->context(), Undefined(env()->isolate()), 1, &uid);
     if (ar.IsEmpty()) {
       ClearFatalExceptionHandlers(env());
       FatalException(env()->isolate(), try_catch);
       return Local<Value>();
     }
   }
 
-  Local<Value> ret = cb->Call(context, argc, argv);
+  prev_id = env()->current_async_id();
+  env()->set_current_async_id(get_uid());
+
+  // Finally... Get to running the user's callback.
+  MaybeLocal<Value> ret = cb->Call(env()->context(), context, argc, argv);
+
+  env()->set_current_async_id(prev_id);
 
-  if (ran_init_callback() && !post_fn.IsEmpty()) {
-    Local<Value> did_throw = Boolean::New(env()->isolate(), ret.IsEmpty());
-    Local<Value> vals[] = { uid, did_throw };
+  Local<Value> ret_v;
+  if (!ret.ToLocal(&ret_v)) {
```

Comment:
I'm not so strong on my C++ V8, what does this mean? I assume it is similar to:
```c++
if (ret.IsEmpty()) {
  return ret;
}
```

but before that was checked/called after the `post` hook.